### ISSUE TITLE
Add braces to multi-line lambda body

### ISSUE THREAD

**Issue by iantabolt** [scalameta/scalafmt#1222]
This issue occurs when you have a single lambda argument with a single statement multi-line body. Scalafmt will drop the lambda to a newline and indent the block. Please correct me if this is working as intended or if a configuration already exists.

With default configuration,

```scala
// Before scalafmt (and expected output)
foos.map(foo =>
  foo
    .withBar("bar")
    .toString)

// After scalafmt
foos.map(
  foo =>
    foo
      .withBar("bar")
      .toString)
```

Meanwhile these similar scenarios are unchanged as expected:

```scala
foos.map(foo => {
  // ...
})

foos.map { foo =>
  // ...
}
```

P.S. for our style guide, I would actually prefer a rewrite rule to add curly braces around multiline lambda bodies:

```scala
// Before 
foos.map(foo =>
  foo
    .withBar("bar")
    .toString)

// After hypothetical rewrite rule
foos.map(foo => {
  foo
    .withBar("bar")
    .toString
})
```

Do you think this is doable/worthwhile? I'm always happy to contribute if you think it is a good addition.

**Comment by olafurpg** [comment]
Thanks for reporting! Moving the argument to the first line is what https://github.com/scalameta/scalafmt/issues/667 is about, and it's been open for a while and my attempts a while ago to fix it failed. I your style guide would prefer curly braces I think you can implement a rewrite rule for this case quite easily so that 
```scala
// Before 
foos.map(foo =>
  foo
    .withBar("bar")
    .toString)

// After hypothetical rewrite rule
foos.map(foo => {
  foo
    .withBar("bar")
    .toString
})
```

Implementing a rewrite rule would also be a good primer for you to rewrite `Foo { ... }` into `Foo({...})` 😉 


Here's a small session from the repl that might get you started

```scala
@ import $ivy.`com.geirsson::scalafmt-core:1.6.0-RC3`, scala.meta._, org.scalafmt._, rewrite._
import $ivy.$                                      , scala.meta._, org.scalafmt._, rewrite._

@ val c = config.ScalafmtConfig()
c: config.ScalafmtConfig = ScalafmtConfig(
  "1.6.0-RC3",
  80,
  ScalaDoc,
  OptIn(true, false, true, true, true, false),
  BinPack(false, false, false, true, 5, List(".*"), List("String", "Term.Name")),
  ContinuationIndent(2, 4, 4),
  Align(
    false,
    false,
    Set(AlignToken("=>", "Case")),
    false,
    true,
    Map(),
    Map(
      "Defn.Trait" -> "class/object/trait",
      "Enumerator.Val" -> "for",
      "Defn.Class" -> "class/object/trait",
      "Defn.Object" -> "class/object/trait",
      "Defn.Val" -> "val/var/def",
      "Defn.Def" -> "val/var/def",
      "Defn.Var" -> "val/var/def",
      "Enumerator.Generator" -> "for"
    )
  ),
  Spaces(Never, false, false, false, List(), true, true, false),
  Literals(Upper, Lower, Lower),
  unix,
...

@ val input = Input.String("""
  object Main {
    lst.map(x =>
      bar.fux(2,
      3)
    )
  }
  """)
input: inputs.Input.String = String(
  """
object Main {
  lst.map(x =>
    bar.fux(2,
    3)
  )
}
"""
)

@ object Curlies extends Rewrite {
    def rewrite(code: Tree, ctx: RewriteCtx): Seq[Patch] = code.collect {
      case Term.Function(_, body) =>
        println(code)
        TokenPatch.AddLeft(body.tokens.head, "{\n", keepTok = true) ::
        TokenPatch.AddRight(body.tokens.last, "\n}", keepTok = true) ::
        Nil
    }.flatten
  }
defined object Curlies

@ Rewrite(input, c.copy(rewrite = c.rewrite.copy(rules = Curlies :: Nil)))

object Main {
  lst.map(x =>
    bar.fux(2,
    3)
  )
}

res21: Input = String(
  """
object Main {
  lst.map(x =>
    {
bar.fux(2,
    3)
}
  )
}
"""
)
```
The tricky part is figuring out the heuristic for when to trigger the rule. You might be able to get away with the simple rule

> insert curly braces when the first first token of the lambda body and the last token are on different lines

```scala
body.tokens.head.pos.startLine < body.tokens.last.pos.endLine
```

**Comment by iantabolt** [comment]
Awesome, I am going to work on making a PR to add this rewrite rule. Since the general issue is kind of redundant with #667, I'll rename this issue to track the rewrite only. Thanks for the huge help getting started! 😃 

Any suggestions on a name? My best shot is `rewrite.redundantBraces.alwaysAroundMultilineLambda = true` but it feels a bit wordy.

**Comment by er1c** [comment]
@iantabolt interesting suggestion, do you often use the `.map( ... )` instead of just `.map{ ... }`?  Our style guide would prefer the rewrite to be something more like:

```
// After hypothetical rewrite rule
foos.map { foo =>
  foo
    .withBar("bar")
    .toString
}
```

**Comment by iantabolt** [comment]
Yeah we use parentheses for all method calls regardless of whether the args are lambdas or values. Our style guide leans towards preferring consistency over all. 

I tried implementing this rewrite rule in #1125 but ended up closing it until I have more time to handle the corner cases. I think it would be a good idea to support your style as well for the rewrite. It's probably more common.

**Comment by kitbellew** [comment]
@iantabolt would fixing #667 amount to addressing this issue? see #1551 

generally, rewriting `(x => y)` to `{ x => y }` might not always be correct scala:

```
class A(a: A => B)

A(a => ...) // correct
A { a => ... } // incorrect
```

also, as @olafurpg mentioned somewhere, rewrite happens before formatting, and an initially multiline body might be reformatted into a single line.

**Comment by iantabolt** [comment]
@kitbellew Yup this issue was originally a duplicate of #667 and fixing that awkward newline was my main goal. That PR looks great!

But to be clear on the braces rewrite, I am suggesting adding curly braces around the _body_ of the lambda, ie

```scala
class A(a: A => B)

A(a => ...) // correct
A(a => { ... }) // still correct?
```

But there are still tricky edge cases where adding braces would compile but change the semantics:

```scala
List(10, 11, 12).map(x =>
  x
  - 5
)
// = List(5, 6, 7)

List(10, 11, 12).map(x => {
  x
  - 5
})
// = List(-5, -5, -5)
```

that bug would exist for outer braces as well

```scala
List(10, 11, 12).map { x =>
  x
  - 5
}
// = List(-5, -5, -5)
```

which is why I kind of gave up on implementing the rewrite

**Comment by kitbellew** [comment]
should be covered by #2019
